### ISSUE TITLE
Empty zip export with DatabaseStorage backend

### ISSUE THREAD

**Issue by jeverling** [kobotoolbox/kobocat#511]
In [PR 453](https://github.com/kobotoolbox/kobocat/pull/453), a change was made to remove temporary files from the zip export creation process. From the description of the related issue #448, it sounds like the assumption was that `NamedTemporaryFiles` would fill up the `/tmp` folder.
However, `NamedTemporaryFiles` should be deleted automatically once the `with` block is exited, at least when there is no exception and `delete=False` wasn't passed: [NamedTemporaryFile](https://docs.python.org/2/library/tempfile.html#tempfile.NamedTemporaryFile)

This change has lead to problems for us. We are using [django-database-files](https://github.com/chrisspen/django-database-files-3000) as storage backend, as one way to achieve scalability.
This backend expects files to be written by a call to `get_storage_class()().save` (as is done to create an empty file before the zip generation).
If you write to the file that `get_storage_class()().open` returns, the changes will end up in a in-memory file, but will not be written to the DB - the storage backend doesn't expect this file to be used for anything but reading. I was hoping to solve this by just adding another call to `get_storage_class()().save`, but this is returning a new file under a different path when the file already exists (which is in accordance to the [spec](https://docs.djangoproject.com/en/2.1/ref/files/storage/#django.core.files.storage.Storage.save)).

So basically we can't do anything else than reverting to using a temporary file again. I realize this is not a problem for the kobotoolbox.org deployment, the boto backend seems to handle writes to the object returned by `get_storage_class()().open` just fine.

Would you still consider merging a pull request that re-introduces temporary files for zip generation? It would also be possible to use a `[Spooled]TemporaryFile` where there never even is a filesystem entry or it's destroyed immediately.

**Comment by jnm** [comment]
@jeverling, I appreciate your contributions very much, but we can't go back to writing ZIP exports to temporary files, at least not when S3 storage is enabled. People collect many, many gigabytes of attachments, and prior to this change, running out of disk space on our application servers was a fairly common occurrence :( I think keeping the entire export in RAM is out of the question as well.

If I'm understanding right, https://github.com/chrisspen/django-database-files-3000 doesn't allow `write()`ing to a file at all, just `save()`ing it in its entirety. That's too bad, but if it helps your use case, why not write up a PR that reintroduces temporary files only when `issubclass(get_storage_class(), DatabaseStorage)` is `True`?
